TELCODOCS-598: filtering out CRs during installation using SiteConfig CRs (#46781)
openshift/openshift-docs: modules/ztp-filtering-ai-crs-using-siteconfig.adoc

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

:_module-type: PROCEDURE
[id="ztp-filtering-ai-crs-using-siteconfig_{context}"]
= Filtering custom resources using SiteConfig filters

Using filters, you can easily customize `SiteConfig` custom resources (CRs) to include or exclude other CRs for use in the installation phase of the zero touch provisioning (ZTP) GitOps pipeline.

The following example configures a custom single-node cluster `SiteConfig` CR that excludes the `siteconfig/extra-manifest/03-sctp-machine-config-worker.yaml` file from the installation.

.Prerequisites

> . You configured the hub cluster for generating the required installation and policy CRs.

Reviewer: Prereqs are an unordered list, not an ordered list. ([procedure](https://raw.githubusercontent.com/redhat-documentation/modular-docs/master/modular-docs-manual/files/TEMPLATE_PROCEDURE_doing-one-procedure.adoc) template)

kind: SiteConfig
metadata:
  name: "site1-sno-du"
  namespace: "site1-sno-du"
spec:
  baseDomain: "example.com"
  pullSecretRef:
    name: "assisted-deployment-pull-secret"
  clusterImageSetNameRef: "openshift-{product-version}"
  sshPublicKey: "<ssh_public_key>" <1>
  clusters:
  - clusterName: "site1-sno-du"
    extraManifestPath: "<custom_manifest_folder>" <2>
    extraManifests: <3>
      filter:
        exclude:
          - 03-sctp-machine-config-worker.yaml
  ...
----
<1> Replace `<ssh_public_key>` with the SSH public key that you use to access the cluster.
<2> Replace `<custom_manifest_folder>` with the name of the `extraManifestPath` folder that contains the custom install CRs, for example `user-custom-manifest/`.
<3> Set the required `include` or `exclude` statements for the install. You can include or exclude any CRs from the folder specified in the `extraManifestPath` field.

. Save the `SiteConfig` CRs  and `/user-custom-manifest` CRs and push them to the site configuration repository.
+
The ZTP pipeline monitors the CRs in the `/user-custom-manifest` directory and adjusts what CRs it applies based on the `SiteConfig` filter instructions.

> . Optional: To remove all CRs at install time including default RAN and user defined CRs, apply the following YAML in the `SiteConfig` CR:

Reviewer: Perhaps
```suggestion
. Optional: To remove all CRs after cluster installation, including default RAN and user-defined CRs, include the following YAML in the `SiteConfig` CR:
```

+
[source,yaml]
----
- clusterName: "site1-sno-du"
  extraManifests:
    filter:
      inclusionDefault: exclude
----

> . Optional: To exclude all extra manifests except for a set of custom CRs, set `inclusionDefault` to `exclude`, add the custom `extraManifestPath` folder location, and specify the `include` files, for example:

Reviewer: > Generally, when you specify the name of a command, parameter, or option in text, specify the word command, parameter, or option after the name.
Command, parameter, and option names are not translated; without the identifying noun, users might be confused about what the name represents.

"directory" is used much more often in the collection than "folder" is.

You define a set of extra manifests for inclusion in the cluster install. These manifests are linked to the `SiteConfig` custom resources (CRs) and are applied to the cluster during installation. Using filters, you can easily customize individual `SiteConfig` CRs based on your requirements.

The following example configures a custom single-node cluster `SiteConfig` CR that excludes `extra-manifest/03-sctp-machine-config-worker.yaml` and instead applies a custom `extra-manifest/custom-sctp-machine-config-worker.yaml` file in the install.

.Prerequisites

. Create a Git repository where you manage your custom site configuration data. The repository must be accessible from the hub cluster and be defined as a source repository for the Argo CD application.

.Procedure

. Create a set of extra manifest CRs that the ZTP pipeline uses to customize the cluster installs.

. Change to your custom `/siteconfig` directory. The following example illustrates a sample `/siteconfig` folder:
+
[source,text]
----
siteconfig
├── site1-sno-du.yaml
└── extra-manifest

>     └── 03-sctp-machine-config-worker.yaml

Reviewer 2: This should not be added to user siteconfig repo. It is part of the source-crs/extra-manifest dir from the ztp repo included in the ztp site generate image by default. 
https://github.com/openshift-kni/cnf-features-deploy/tree/master/ztp/source-crs/extra-manifest

Reviewer 3: yes, only the `custom-sctp-machine-config-worker.yaml` is needed

Author: Updated. I wasn't sure about `- cluster:` from https://github.com/openshift-kni/cnf-features-deploy/blob/master/ztp/gitops-subscriptions/argocd/ExtraManifestsFilter.md example  since it is not in http://registry.kni-qe-0.lab.eng.rdu2.redhat.com:3000/kni-qe/ztp-site-configs/src/worker-2-4.10/siteconfig/worker-2.yaml#L33-L36

│            └── 03-sctp-machine-config-worker.yaml
└── user-custom-manifest
      └── custom-sctp-machine-config-worker.yaml
----

. Edit the custom `site1-sno-du.yaml` file to set the `exclude` files, for example:
+
[source,yaml]
----
apiVersion: ran.openshift.io/v1
kind: SiteConfig
metadata:
  name: "site1-sno-du"
  namespace: "site1-sno-du"
spec:
  baseDomain: "example.com"
  pullSecretRef:
    name: "assisted-deployment-pull-secret"
  clusterImageSetNameRef: "openshift-4.10"
  sshPublicKey: "<ssh_public_key>"
  clusters:
  - clusterName: "site1-sno-du"
    networkType: "OVNKubernetes"
    clusterLabels:
      sites : "site1-sno-du"

    ...

    extraManifestPath:
    extraManifests: <1>
      filter:

>         exclude:

Reviewer 2: Perhaps we should mention which CRs are exclusible? Seems it's easy to cause confusion unless user is familiar with how ztp works internally.

Author: "You can include or exclude any CRs from the folder specified in the `extraManifestPath` field."?

Reviewer 2: The exclusible CRs are usually the ones built into the ztp site generator image under source-crs/extra-manifest/ dir. 
Although it can also exclude custom CRs customer added, but it is not a common use case, because it would have been easier to not add them at all. 
I'm not sure whats the best way to describe this. 

@vitus133 could you please suggest the wording?

Reviewer 3: I would suggest "You can exclude any CRs from "source-crs/extra-manifest", which are otherwise included by default". Or something like that. To me it wouldn't make sense to add a CR to the extraManifestPath and then exclude it in the siteconfig

Author: I added this clarification to the procedure preamble

>   extraManifestPath: extra-manifest <1>

Reviewer 2: Perhaps a different name than extra-manifest would be ideal. I'm not sure if this would cause any problem when the user manifests dir is having the same name as ztp source-crs/manifest-dir. But it may cause confusion even if it will be properly handled.

Another small thing is qe has been using "my-manifests-dir/" format as described in README. Not sure if "my-manifests-dir" would also work (please confirm with SME).

>       include:

Reviewer 2: I would suggest to use the exact same/very similar example(s) as here:
https://github.com/openshift-kni/cnf-features-deploy/blob/master/ztp/gitops-subscriptions/argocd/ExtraManifestsFilter.md 

In current example in this doc, the "include" is redundant because all files under extra-manifest/ dir would have been added by default. "include" is only needed when ALL the extraManifests are excluded like in the ExtraManifestsFilter.md.

Author: updated. PTAL
